main.c

```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
     if (selector != 0) {
         fw_cfg_get(selector, &params.cmdline_addr, sizeof(params.cmdline_addr));
     }
-
     return params;
 }
 
-int main(void)
+// Linux boot follows the Linux x86 32-bit Boot Protocol
+// (https://www.kernel.org/doc/html/latest/x86/boot.html#bit-boot-protocol).
+void linux_boot(struct boot_params boot_params)
 {
-    struct bzimage_params bzimage_params = initialize_from_fw_cfg();
-
-    die_on(bzimage_params.bzimage_addr == 0, "bzImage start address not found!\n"
-           "The VMM does not offer a Linux kernel via fw-cfg. Is the relevant config option missing?\n");
+    die_on(
+        boot_params.kernel_addr == 0,
+        "bzImage start address not found!\n"
+        "The VMM does not offer a Linux kernel via fw-cfg. Is the relevant config option missing?\n");
 
-    printf("Loading Linux kernel from address: 0x%lx\n", bzimage_params.bzimage_addr);
+    printf("Loading Linux kernel from address: 0x%lx\n", boot_params.kernel_addr);
 
-    char *linux_header_end = (char *)(bzimage_params.bzimage_addr + 0x201);
+    char *linux_header_end = (char *)(boot_params.kernel_addr + 0x201);
     size_t linux_header_size = 0x202 + *linux_header_end - LINUX_HEADER_OFFSET;
 
-    struct linux_params *boot_params = malloc(sizeof(*boot_params));
-    memset(boot_params, 0, sizeof(*boot_params));
+    struct linux_params *linux_params = malloc(sizeof(*linux_params));
+    memset(linux_params, 0, sizeof(*linux_params));
 
-    die_on(LINUX_HEADER_OFFSET + linux_header_size > sizeof(*boot_params),
+    die_on(LINUX_HEADER_OFFSET + linux_header_size > sizeof(*linux_params),
            "Invalid linux header size");
 
-    memcpy((char *)boot_params + LINUX_HEADER_OFFSET,
-           (char *)(bzimage_params.bzimage_addr + LINUX_HEADER_OFFSET),
+    memcpy((char *)linux_params + LINUX_HEADER_OFFSET,
+           (char *)(boot_params.kernel_addr + LINUX_HEADER_OFFSET),
```

Comment:
nit: While you are here, can you make the source operands of `memcpy` `const` pointers?

Reply:
Fixed.